src/index.ts

```diff
@@ -81,24 +81,26 @@ export class VuexPersistence<S> implements PersistOptions<S> {
      * How this works is -
      *  1. If there is options.reducer function, we use that, if not;
      *  2. We check options.modules;
-     *    1. If there is no options.modules array, we use entire state in reducer
-     *    2. Otherwise, we create a reducer that merges all those state modules that are
+     *    1a. If there is no options.modules array, we use entire state in reducer
+     *    1b. If writes are async, we clone the state so each saveState call in the queue has its
+     *        own caopy of the state.
+     *    2.  Otherwise, we create a reducer that merges all those state modules that are
      *        defined in the options.modules[] array
      * @type {((state: S) => {}) | ((state: S) => S) | ((state: any) => {})}
      */
-    this.reducer = (
-      (options.reducer != null)
-        ? options.reducer
-        : (
-          (options.modules == null)
-            ? ((state: S) => state)
-            : (
-              (state: any) =>
-                (options!.modules as string[]).reduce((a, i) =>
-                  merge(a, { [i]: state[i] }, this.mergeOption), {/* start empty accumulator*/ })
-            )
-        )
-    )
+    if (options.reducer != null) {
+      this.reducer = options.reducer
+    } else if (options.modules == null) {
+      if (options.asyncStorage) {
+        this.reducer = (state: S) => merge({}, state, this.mergeOption)
+      } else {
+        this.reducer = (state: S) => state
```

Comment:
This addresses an issue that only occurs with async storage. The vuex store is mutable, so we need to create a snapshot of the store for each copy of `saveState`. Without this code, if we mutate the store a few times in a row, the first call to `saveState` may end up seeing all the changes, when it should really only see the state of the store after the first write.

Reply:
But our code never executes this part right? Because we always have a reduce set: https://github.com/propelinc/freshebt-vue/blob/98bf640ab664ddc1a4448e1b7b5048cf12a45d7b/src/store/index.ts#L233

so are we just fixing for other folks in the future?

Reply:
I think this also possibly leaves the bug around if the provided reducers do not themselves deeply copy objects. From my reading of our reducer in freshebt-server, ours does not seem to do this. One way to fix this would be to wrap any provided reducer in a function that returns a copy (perhaps only when the store is async). Another way would be to just change the semantics so that the reducer should already return the reduced state in JSON format. I would probably choose the former, since the later probably requires a lot more refactoring.

Reply:
@rammie I reverted this and we now clone the state before calling `saveState` in the async case.